### PR TITLE
Fix AOT warnings for RuntimeContext

### PR DESCRIPTION
**Pull request by utpilla**
Towards #3429 

This PR introduces a **Breaking change**. This would no longer support assigning custom types to `RuntimeContext.ContextSlotType`.

## Changes
- Fix AOT warnings for `RuntimeContext` class
- Drop support for custom implementations of `RuntimeContextSlot<T>`

## Merge requirement checklist
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial changes

### CODE REVIEW

**Review by reyang — APPROVED** [review]

**Review by Yun-Ting — APPROVED** [review]

**Review by eerhardt — APPROVED** [review]
Looks great! Thanks!

File: src/OpenTelemetry.Api/Context/RuntimeContext.cs [hunk 1]
```diff
@@ -27,10 +27,40 @@ public static class RuntimeContext
...
+#if NETFRAMEWORK
+                else if (value == typeof(RemotingRuntimeContextSlot<>))
+                {
+                    contextSlotType = value;
+                }
+#endif
+                else
+                {
+                    throw new NotSupportedException($"{value} is not a supported type.");
+                }
```
Comment (Kielek) [inline src/OpenTelemetry.Api/Context/RuntimeContext.cs:59]: Should it really throw Exception? As I remember OTel SDK/API should rather silently ignore non-working parameters than throwing exception.

Just to double check if Event/Log is not a better option.

File: src/OpenTelemetry.Api/CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,13 @@
 
 ## Unreleased
 
+* **Breaking change** `RuntimeContext.ContextSlotType` can only be assigned one
+  of the following types: `AsyncLocalRuntimeContextSlot<>`,
+  `ThreadLocalRuntimeContextSlot<>`, and `RemotingRuntimeContextSlot<>`. A
+  `System.NotSupportedException` will be thrown if you try to assign any type
+  other than the three types mentioned.
+  ([#4542](https://github.com/open-telemetry/opentelemetry-dotnet/pull/4542))
```
Comment (alanwest) [inline src/OpenTelemetry.Api/CHANGELOG.md:10]:
```suggestion
* **Breaking change** In order to make `RuntimeContext` compatible with
  ahead-of-time compilation (AOT),
  `RuntimeContext.ContextSlotType` can only be assigned one
  of the following types: `AsyncLocalRuntimeContextSlot<>`,
  `ThreadLocalRuntimeContextSlot<>`, and `RemotingRuntimeContextSlot<>`. A
  `System.NotSupportedException` will be thrown if you try to assign any type
  other than the three types mentioned.
  ([#4542](https://github.com/open-telemetry/opentelemetry-dotnet/pull/4542))
```

**Review by alanwest — APPROVED** [review]
Just a suggestion to add context for why this change is being made.

File: src/OpenTelemetry.Api/Context/RuntimeContext.cs (hunk 1 above)
Comment (eerhardt) [reply src/OpenTelemetry.Api/Context/RuntimeContext.cs:59]: In this case, this isn't a "someone configured it wrong" or "something went wrong" situation. This is a "the developer used the API incorrectly" situation. So I think it is better to let the developer know early that what they are trying to do isn't possible. (i.e. throw an exception).

Comment (reyang) [reply src/OpenTelemetry.Api/Context/RuntimeContext.cs:59]: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/error-handling.md#basic-error-handling-principles

_The API or SDK MAY fail fast and cause the application to fail on initialization_

**Pull request merged**